01-algorithms/01-algorithms.md

```diff
@@ -50,65 +61,454 @@ The study of computer science is the study of algorithms.  Researchers in CS foc
+
+### Explore Example Input/Output
+
+Next to ensure we understand the problem we need to develop example inputs and determine what the output *should* be for those given inputs.  In industry this might involve drawing mock-ups of the program and running through scenarios or asking a customer or a team member to do so.  It could also involve generating sample input data and determining what the output should be.
+
+While we want to ensure our algorithm will work on typical input or **nominal cases**, it is also very important to develop a few examples which will test the boundaries of what is possible for input.  These are called **edge cases**.
+
+We should consider:
+* **Truthy Nominal Case**
+  * In this example we could give a list that contains one integer that occurs twice within the given distance `k`.
+  * For example: an input list of `[1, 0, 2, 1, 3, 4]` and `k` value 5 would return `True` because the integer 1 occurs twice: once at index 0 and once at index 3 which is within 5 indices.
+* **Falsy Nominal Case**
+  * In this example we could give a list where no integers occur twice within the given distance. 
+  * For example: an input list of `[1, 0, 2, 1, 3, 4]` and `k` value 1 would return `False` because while 1 occurs twice in the given list, their distance from each other is greater than 1
+* **Input List Edge Case**
+  * In this example we could give an input list on the edge of what is possible. 
+    * For example: an empty input list `[]` and any value of `k` would need to return *something* to indicate that it is not possible for there to be duplicates when the list is empty.  
+    * Another example: an input list of `[1]` and any value of `k` needs to return *something* to indicate that it is not possible for there to be duplicates when there is only one element in the list.
+* **K Edge Case**
+  * This would be an input on the edge of what is possible for our integer value `k`
+  * For example: given any input list and a `k` value of 0 would need to return *something* to indicating either that `k` must be greater than one or that it is not possible for there to be duplicates within 0 distance of each other (because 0 distance means we can only be looking at a single element)
+  
+By looking at concrete sample input and output examples, we can see that our problem is not as simple as it seems. We need to consider cases where we have a list with less than two elements or `k` is zero. 
+
+For our purposes we will expect the function to return an empty list `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.
```

Comment:
```suggestion
For our purposes we will expect the function to return `False` if the length of the list is less than two or if `k` is zero. In an interview setting, consider asking your interviewer if there is a preferred return value when the algorithm encounteres the edge case scenearios. If there is no preferred return value, make a thoughtful decision about what the algorithm should return and communicate your choice to your interviewer.
```

Not sure if this was a typo. Feel free to ignore the suggestion if I misread!

Reply:
Nice catch, thank you!